src/btree/bt_cursor.c

```diff
@@ -232,107 +232,195 @@ __cursor_fix_implicit(WT_BTREE *btree, WT_CURSOR_BTREE *cbt)
-    upd = NULL;
 
     /*
-     * We may be pointing to an insert object, and we may have a page with
-     * existing entries.  Insert objects always have associated update
-     * objects (the value).  Any update object may be deleted, or invisible
-     * to us.  In the case of an on-page entry, there is by definition a
-     * value that is visible to us, the original page cell.
-     *
-     * If we find a visible update structure, return our caller a reference
-     * to it because we don't want to repeatedly search for the update, it
-     * might suddenly become invisible (imagine a read-uncommitted session
-     * with another session's aborted insert), and we don't want to handle
-     * that potential error every time we look at the value.
+     * Check for a value on disk or in the history store, passing in any update.
      *
-     * Unfortunately, the objects we might have and their relationships are
-     * different for the underlying page types.
+     * Potentially checking an update chain twice (both here, and above if the insert list is set in
+     * the case of column-store), isn't a mistake. In a modify chain, if rollback-to-stable recovers
+     * a history store record into the update list, the base update may be in the update list and we
+     * must use it rather than falling back to the on-disk value as the base update.
+     */
+    WT_RET(__wt_txn_read(session, cbt, key, cbt->recno, upd));
+    if (cbt->upd_value->type != WT_UPDATE_INVALID) {
+        if (cbt->upd_value->type == WT_UPDATE_TOMBSTONE)
+            return (0);
+        *valid = true;
+    }
+    return (0);
+}
+
+/*
+ * __cursor_valid_row --
+ *     Determine if a valid update is being referenced by the cursor for row-store.
+ */
+static int
+__cursor_valid_row(WT_CURSOR_BTREE *cbt, bool *valid, bool check_bounds)
+{
+    WT_DECL_ITEM(key);
+    WT_PAGE *page;
+    WT_SESSION_IMPL *session;
+    WT_UPDATE *upd;
+    bool key_out_of_bounds;
+
+    session = CUR2S(cbt);
+    upd = NULL;
+    page = cbt->ref->page;
+    key_out_of_bounds = false;
+
+    /* For all row-store cases we can extract the key from cbt->tmp if the compare value is zero. */
+    if (cbt->compare == 0 && cbt->tmp != NULL)
+        key = cbt->tmp;
+
+    /*
+     * In the case of row-store, an insert object implies ignoring any page objects, no insert
+     * object can have the same key as an on-page object.
      *
-     * In the case of row-store, an insert object implies ignoring any page
-     * objects, no insert object can have the same key as an on-page object.
      * For row-store:
      *	if there's an insert object:
      *		if there's a visible update:
      *			exact match
      *		else
      *			no exact match
      *	else
-     *		use the on-page object (which may have an associated
-     *		update object that may or may not be visible to us).
+     *		use the on-page object (which may have an associated update object that may or may
+     *not be visible to us).
```

Comment:
Nitpick, it'd probably make more sense to have the `not` line up with the line above, kinda like how it was before.

Reply:
42809baf4